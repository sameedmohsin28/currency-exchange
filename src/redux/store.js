import { configureStore } from '@reduxjs/toolkit';
import currencySlice from './currency/currencySlice';

const store = configureStore({
  reducer: currencySlice.reducer,
});

export default store;
