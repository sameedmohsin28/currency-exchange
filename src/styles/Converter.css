.Converter {
  background-color: rgb(32, 32, 97);
  min-height: 100vh;
}

.back-button {
  color: white;
  text-decoration: none;
}

.converter-section {
  display: flex;
  flex-direction: column;
  gap: 40px;
  align-items: center;
  justify-content: center;
  margin-top: 20px;
}

.exchange-rate-text {
  margin: 0;
  color: white;
}

.base-to-specific,
.specific-to-base {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;
}

.base-to-specific-heading,
.specific-to-base-heading {
  margin: 0;
  color: white;
}

.amount,
.converted {
  height: 30px;
  width: 220px;
  border: 5px solid white;
  background-color: rgb(47, 41, 53);
  color: white;
  padding: 10px;
}

.amount::placeholder,
.converted::placeholder {
  color: white;
  opacity: 0.5;
}

.amount::-webkit-outer-spin-button,
.amount::-webkit-inner-spin-button,
.converted::-webkit-outer-spin-button,
.converted::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.amount:focus,
.converted:focus {
  outline: 0;
}
