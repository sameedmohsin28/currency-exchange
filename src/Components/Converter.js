import { useEffect, useState } from 'react';
import { useLocation, Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import '../styles/Converter.css';
import logo from '../assets/logoForWhite.png';
import { fetchPreviousRates } from '../redux/api';

const Converter = () => {
  const dispatch = useDispatch();
  const location = useLocation();
  const { rate } = location.state;
  const { symbol } = location.state;
  const { timeSeries } = location.state;

  const { isLoadingTimeSeriesArray } = useSelector((store) => (store));
  const { timeSeriesArray } = useSelector((store) => (store));
  const reversedTimeSeriesArray = [...timeSeriesArray].reverse();

  useEffect(() => {
    dispatch(fetchPreviousRates(timeSeries, symbol));
  }, [dispatch, timeSeries, symbol]);

  const [answer, setAnswer] = useState();
  const [question, setQuestion] = useState();

  const calculateAnswer = (e) => {
    setAnswer(e.target.value * rate);
    if (!e.target.value) {
      setAnswer('');
    }
  };

  const calculateQuestion = (e) => {
    setQuestion(e.target.value / rate);
    if (!e.target.value) {
      setQuestion('');
    }
  };

  if (isLoadingTimeSeriesArray) {
    return (
      <>
        <div className="Converter">
          <header className="header">
            <Link to="/" className="logo-image-and-name">
              <img src={logo} alt="logo" className="logo-image" />
              <h1 className="logo-name">ForEx Rates</h1>
            </Link>
            <div>
              <Link to="/" className="back-button">
                Back
              </Link>
            </div>
          </header>
          <section className="details-section">
            <div className="converter-div">
              <div className="exchange-rate">
                <h2 className="exchange-rate-text">
                  {'1 USD = '}
                  {rate}
                  {' '}
                  {symbol}
                </h2>
              </div>
              <div className="base-to-specific">
                <h3 className="base-to-specific-heading">
                  {'USD to '}
                  {symbol}
                </h3>
                <input type="number" placeholder="Enter Amount in USD" onChange={calculateAnswer} className="amount" />
                <input type="number" placeholder={`Converted Amount in ${symbol}`} value={answer} className="converted" />
              </div>
              <div className="specific-to-base">
                <h3 className="specific-to-base-heading">
                  {symbol}
                  {' to USD'}
                </h3>
                <input type="number" placeholder={`Enter Amount in ${symbol}`} onChange={calculateQuestion} className="amount" />
                <input type="number" placeholder="Converted Amount in USD" value={question} className="converted" />
              </div>
            </div>
            <div className="isLoading-state">
              <h3 className="isLoading-message">Fetching rates from past 10 days...</h3>
            </div>
          </section>
        </div>
      </>
    );
  }

  return (
    <>
      <div className="Converter">
        <header className="header">
          <Link to="/" className="logo-image-and-name">
            <img src={logo} alt="logo" className="logo-image" />
            <h1 className="logo-name">ForEx Rates</h1>
          </Link>
          <div>
            <Link to="/" className="back-button">
              Back
            </Link>
          </div>
        </header>
        <section className="details-section">
          <div className="converter-div">
            <div className="exchange-rate">
              <h2 className="exchange-rate-text">
                {'1 USD = '}
                {rate}
                {' '}
                {symbol}
              </h2>
            </div>
            <div className="base-to-specific">
              <h3 className="base-to-specific-heading">
                {'USD to '}
                {symbol}
              </h3>
              <input type="number" placeholder="Enter Amount in USD" onChange={calculateAnswer} className="amount" />
              <input type="number" placeholder={`Converted Amount in ${symbol}`} value={answer} className="converted" />
            </div>
            <div className="specific-to-base">
              <h3 className="specific-to-base-heading">
                {symbol}
                {' to USD'}
              </h3>
              <input type="number" placeholder={`Enter Amount in ${symbol}`} onChange={calculateQuestion} className="amount" />
              <input type="number" placeholder="Converted Amount in USD" value={question} className="converted" />
            </div>
          </div>
          <div className="past-data-mesaage-div">
            <h4 className="past-data-message">Conversion rates from past 10 days</h4>
          </div>
          <div className="past-data-div">
            {reversedTimeSeriesArray.map((pastRate) => (
              <div key={pastRate.date} className="each-past-data">
                <p className="pastRate-date">{pastRate.date}</p>
                <p className="pastRate-rate">{pastRate.rateAmount}</p>
              </div>
            ))}
          </div>
        </section>
      </div>
    </>
  );
};

export default Converter;
