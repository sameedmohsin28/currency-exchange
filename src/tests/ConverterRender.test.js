import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import store from '../Redux/store';

import Converter from '../Components/Converter';

describe('Converter rendering Validation', () => {
  test('Converter should render with / url', () => {
    const treeSnapShot = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route path="/converter" element={<Converter />} />
          </Routes>
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(treeSnapShot).toMatchSnapshot();
  });
});