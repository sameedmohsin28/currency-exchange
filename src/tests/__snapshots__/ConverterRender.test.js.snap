// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Converter rendering Validation Converter should render with / url 1`] = `null`;
